File: 5 Testen und Debugging/(1) Digitaluhr_V1/WeckStoppUhr.java
/**
 * Die Klasse WeckStoppUhr simuliert eine Digitaluhr, 
 * ... mit Stoppuhr- und Weckfunktion. 
 * Autor:   Martin Lentz
 * Version: 01.02.2017
 */
public class WeckStoppUhr
{
    private Uhrenanzeige uhr, wecker, stoppuhr;
    private boolean weckerEingeschaltet, alarm, stoppuhrLaeuft;

    public WeckStoppUhr()
    {
        wecker = new Uhrenanzeige();
        stoppuhr = new Uhrenanzeige();
        uhr = new Uhrenanzeige();
        weckerEingeschaltet = alarm = stoppuhrLaeuft = false;
    }
    
    // Erhöht die Zeit um einen Takt, d. h. um eine Sekunde 
    public void erhoeheZeit()
    {
        uhr.erhoeheZeit();

        if (stoppuhrLaeuft) stoppuhr.erhoeheZeit();

        if (weckerEingeschaltet) wecker_pruefeWeckzeit();
    }

    // Stellt neue Uhrzeit ein
    public void uhr_setzeUhrzeit(int stunden, int minuten, int sekunden)
    {
        uhr.setzeZeit(stunden, minuten, sekunden);
    }

    // Liefert die aktuelle Uhrzeit 
    public String uhr_gibUhrzeit()
    {
        return uhr.gibZeit();
    }    
    
    // Schaltet den Wecker ein
    public void wecker_einschalten()
    {
        weckerEingeschaltet = true;
    }
        
    // Schaltet den Wecker aus
    public void wecker_ausschalten()
    {
        weckerEingeschaltet = false;
        wecker_AlarmAusschalten();
    }
        
    // Schaltet den Alarm aus ohne den Wecker auszuschalten
    public void wecker_AlarmAusschalten()
    {
        alarm = false;
    }
        
    // Schlummerfunktion: Deaktiviert den Alarm für 10 Minuten
    public void wecker_schlummern()
    {
        if (alarm)
        {
            alarm = false;
            for (int i = 0; i < 600; i++)
            {
                wecker.erhoeheZeit();
            }
        }
    }
    
    // Stellt die Weckzeit ein
    public void wecker_setzeWeckzeit(int stunden, int minuten, int sekunden)
    {
        wecker.setzeZeit(stunden, minuten, sekunden);
    }    

    // Liefert die eingestellte Weckzeit 
    public String wecker_gibWeckzeit()
    {
        return wecker.gibZeit();
    }    

    // Prueft, ob die Weckzeit erreicht ist und aktiviert dann den Alarm
    private void wecker_pruefeWeckzeit()
    {
        if (wecker_gibWeckzeit().equals(uhr_gibUhrzeit())) alarm = true;
    }

    // Liefert den Aktivierungszustand des Weckers
    public boolean wecker_gibWeckerEingeschaltet()
    {
         return weckerEingeschaltet;
    }
    
    // Liefert den Alarm
    public boolean wecker_gibAlarm()
    {
        return alarm;
    }

    // Startet die Stoppuhr
    public void stoppuhr_starten()
    {
        stoppuhrLaeuft = true;
    }
    
    // Stoppt die Stoppuhr
    public void stoppuhr_stoppen()
    {
        stoppuhrLaeuft = false;
    }
    
    // Stellt die Stoppuhr zurück auf Null
    public void stoppuhr_reset()
    {
        stoppuhr_stoppen();
        stoppuhr.setzeZeit(0, 0, 0);
    }

    // Liefert die aktuelle Stoppuhrzeit
    public String stoppuhr_gibStoppuhrzeit()
    {
        return stoppuhr.gibZeit();
    }
    
    // Liefert den Aktivierungszustand der Stoppuhr
    public boolean stoppuhr_gibStoppuhrLaeuft()
    {
        return stoppuhrLaeuft;
    }
}

File: Informatik/2 C#-Grundlagen und GUIs/8 Digitaluhr_V2/Digitaluhr_V2/WeckStoppUhr.cs
namespace Digitaluhr_V2
{
    public class WeckStoppUhr
    {
        private readonly Uhrenanzeige stoppuhr;
        private readonly Uhrenanzeige uhr;
        private readonly Uhrenanzeige wecker;
        
        private int schlummernAnzahl;
        private bool weckerEingeschaltet, alarm, stoppuhrLaeuft;

        public WeckStoppUhr()
        {
            wecker = new Uhrenanzeige();
            stoppuhr = new Uhrenanzeige();
            uhr = new Uhrenanzeige();

            weckerEingeschaltet = false;
            alarm = false;
            stoppuhrLaeuft = false;

            schlummernAnzahl = 0;
        }

        // Main Loop
        public void ErhoeheZeit()
        {
            uhr.ErhoeheZeit();
            if (stoppuhrLaeuft) stoppuhr.ErhoeheZeit();
            if (weckerEingeschaltet) Wecker_pruefeWeckzeit();
        }

        // Uhr
        public void Uhr_setzeUhrzeit(int stunden, int minuten, int sekunden)
        {
            uhr.SetzeZeit(stunden, minuten, sekunden);
        }

        public string Uhr_gibUhrzeit()
        {
            return uhr.ToString();
        }

        // Wecker
        public void Wecker_einschalten()
        {
            weckerEingeschaltet = true;
        }

        public void Wecker_ausschalten()
        {
            weckerEingeschaltet = false;
            wecker_AlarmAusschalten();
        }

        public void wecker_AlarmAusschalten()
        {
            alarm = false;
            for (var i = 0; i < 24 * 60 * 60 - schlummernAnzahl * 300; i++) wecker.ErhoeheZeit();
            schlummernAnzahl = 0;
        }

        public void Wecker_schlummern()
        {
            if (alarm)
            {
                alarm = false;
                schlummernAnzahl++;
                for (var i = 0; i < 300; i++) wecker.ErhoeheZeit();
            }
        }

        public void Wecker_setzeWeckzeit(int stunden, int minuten, int sekunden)
        {
            wecker.SetzeZeit(stunden, minuten, sekunden);
        }

        public string Wecker_gibWeckzeit()
        {
            return wecker.ToString();
        }

        private void Wecker_pruefeWeckzeit()
        {
            if (Wecker_gibWeckzeit() == Uhr_gibUhrzeit()) alarm = true;
        }

        public bool Wecker_gibWeckerEingeschaltet()
        {
            return weckerEingeschaltet;
        }

        public bool Wecker_gibAlarm()
        {
            return alarm;
        }

        // Stoppuhr
        public void Stoppuhr_starten()
        {
            stoppuhrLaeuft = true;
        }

        public void Stoppuhr_stoppen()
        {
            stoppuhrLaeuft = false;
        }

        public void Stoppuhr_reset()
        {
            Stoppuhr_stoppen();
            stoppuhr.SetzeZeit(0, 0, 0);
        }

        public string Stoppuhr_gibStoppuhrzeit()
        {
            return stoppuhr.ToString();
        }

        public bool Stoppuhr_gibStoppuhrLaeuft()
        {
            return stoppuhrLaeuft;
        }
    }
}